File: clients/ios/PandaSolve/Data/TaskRepository.swift
import Foundation

final class TaskRepository {
    private let api: APIClient
    private var taskCache: [String: TaskDetail] = [:]
    init(api: APIClient) { self.api = api }

    func submitImage(data: Data, caption: String?) async throws -> String {
        var fields: [String: String] = [:]
        if let caption { fields["caption"] = caption }
        let resp: TaskRef = try await api.postMultipart(
            "/v1/tasks",
            fields: fields,
            file: (name: "file", filename: "task.jpg", data: data, mime: "image/jpeg")
        )
        return resp.taskId
    }

    func submitText(text: String) async throws -> String {
        let resp: TaskRef = try await api.post("/v1/tasks/text", body: TaskCreateText(text: text))
        return resp.taskId
    }

    func cachedTask(_ id: String) -> TaskDetail? { taskCache[id] }

    func get(id: String) async throws -> TaskDetail {
        let t: TaskDetail = try await api.get("/v1/tasks/\(id)")
        taskCache[id] = t
        return t
    }

    func list(limit: Int, before: String?, albumId: String? = nil, q: String? = nil) async throws -> TaskList {
        var query: [URLQueryItem] = [URLQueryItem(name: "limit", value: "\(limit)")]
        if let before { query.append(URLQueryItem(name: "before", value: before)) }
        if let albumId { query.append(URLQueryItem(name: "album_id", value: albumId)) }
        if let q, !q.isEmpty { query.append(URLQueryItem(name: "q", value: q)) }
        return try await api.get("/v1/tasks", query: query)
    }

    func rename(id: String, title: String) async throws -> TaskDetail {
        let t: TaskDetail = try await api.patch("/v1/tasks/\(id)", body: TaskUpdateRequest(title: title))
        taskCache[id] = t
        return t
    }

    func chatHistory(id: String) async throws -> ChatThread {
        try await api.get("/v1/tasks/\(id)/chat")
    }

    func sendChat(id: String, message: String) async throws -> ChatThread {
        try await api.post("/v1/tasks/\(id)/chat", body: ChatSendRequest(message: message))
    }

    /// Follow-up question with an attached photo (vision context) + caption.
    func sendChatImage(id: String, data: Data, caption: String) async throws -> ChatThread {
        try await api.postMultipart(
            "/v1/tasks/\(id)/chat/image",
            fields: ["message": caption],
            file: (name: "file", filename: "attach.jpg", data: data, mime: "image/jpeg"))
    }
}

File: clients/android/app/src/main/java/com/pandasolve/app/data/repository/TaskRepository.kt
package com.pandasolve.app.data.repository

import com.pandasolve.app.domain.model.ChatSendRequest
import com.pandasolve.app.domain.model.ChatThread
import com.pandasolve.app.domain.model.TaskCreateText
import com.pandasolve.app.domain.model.TaskDetail
import com.pandasolve.app.domain.model.TaskList
import com.pandasolve.app.domain.model.TaskUpdateRequest
import com.pandasolve.app.network.PandaApiService
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody

@Singleton
class TaskRepository @Inject constructor(
    private val api: PandaApiService,
) {
    suspend fun submitImage(bytes: ByteArray, caption: String?): String {
        val filePart = MultipartBody.Part.createFormData(
            name = "file",
            filename = "task.jpg",
            body = bytes.toRequestBody("image/jpeg".toMediaTypeOrNull()),
        )
        val captionPart = caption?.toRequestBody("text/plain".toMediaTypeOrNull())
        return api.submitImage(filePart, captionPart).taskId
    }

    suspend fun submitText(text: String): String =
        api.submitText(TaskCreateText(text = text)).taskId

    // Cache loaded tasks so re-opening one paints instantly while it refreshes.
    private val taskCache = ConcurrentHashMap<String, TaskDetail>()
    fun cachedTask(id: String): TaskDetail? = taskCache[id]

    suspend fun get(id: String): TaskDetail = api.getTask(id).also { taskCache[id] = it }

    suspend fun rename(id: String, title: String): TaskDetail =
        api.updateTask(id, TaskUpdateRequest(title = title)).also { taskCache[id] = it }

    suspend fun chatHistory(taskId: String): ChatThread = api.getChat(taskId)

    suspend fun sendChatImage(taskId: String, bytes: ByteArray, caption: String): ChatThread {
        val filePart = MultipartBody.Part.createFormData(
            name = "file", filename = "chat.jpg",
            body = bytes.toRequestBody("image/jpeg".toMediaTypeOrNull()),
        )
        val msgPart = caption.toRequestBody("text/plain".toMediaTypeOrNull())
        return api.postChatImage(taskId, filePart, msgPart)
    }

    suspend fun sendChat(taskId: String, message: String): ChatThread =
        api.postChat(taskId, ChatSendRequest(message))

    suspend fun list(limit: Int, before: String?, albumId: String? = null, q: String? = null): TaskList =
        api.listTasks(limit = limit, before = before, albumId = albumId, q = q?.ifBlank { null })
}
